File: code_combat/8_Cloudrip_Mountain/444-Reaping_Fire/reaping_fire.py
def chooseStrategy():
    enemies = hero.findEnemies()
    if hero.gold >= hero.costOf("griffin-rider"):
        return "griffin-rider"
    for enemy in enemies:
        if enemy.type == "fangrider" and enemy.pos.x < 55:
            return "fight-back"
    return "collect-coins"


def commandAttack():
    friends = hero.findFriends()
    for friend in friends:
        enemy = friend.findNearestEnemy()
        if enemy:
            hero.command(friend, "attack", enemy)


def pickUpCoin():
    item = hero.findNearestItem()
    if (item)
        hero.move(item.pos)


def heroAttack():
    enemy = hero.findNearestEnemy()
    if enemy:
        if hero.canCast("regen"):
            hero.cast("regen", hero)
        while enemy.health > 0:
            hero.attack(enemy)


while True:
    if hero.time > 55:
        if hero.isReady("reset-cooldown"):
            hero.resetCooldown("invisibility")
        if hero.canCast("invisibility"):
            hero.cast("invisibility", hero)
            hero.moveXY(100, 20)
        continue
    if hero.time > 50:
        if hero.canCast("invisibility"):
            hero.cast("invisibility", hero)
            hero.moveXY(100, 20)
    commandAttack()
    strategy = chooseStrategy()
    if strategy == "griffin-rider":
        hero.summon("griffin-rider")
    elif strategy == "fight-back":
        heroAttack()
    else:
        pickUpCoin()

File: projects/code_combat/8_Cloudrip_Mountain/444-Reaping_Fire/reapingFire.js
// The goal is to survive for 30 seconds, and keep the mines intact for at least 30 seconds.

function chooseStrategy() {
    var enemies = hero.findEnemies();
    // If you can summon a griffin-rider, return "griffin-rider"
    if (hero.gold >= hero.costOf("griffin-rider"))
        return "griffin-rider";
    // If there is a fangrider on your side of the mines, return "fight-back"
    for (var i = 0; i < enemies.length; i++)
        if (enemies[i].type == "fangrider" && enemies[i].pos.x < 55)
            return "fight-back";
    // Otherwise, return "collect-coins"
    return "collect-coins";
}

function commandAttack() {
    // Command your griffin riders to attack ogres.
    var friends = hero.findFriends();
    for (var i = 0; i < friends.length; i++) {
        var friend = friends[i];
        var enemy = friend.findNearestEnemy();
        if (enemy)
            hero.command(friend, "attack", enemy);
    }
}

function pickUpCoin() {
    // Collect coins
    var item = hero.findNearestItem();
    if (item)
        hero.move(item.pos);
}

function heroAttack() {
    // Your hero should attack fang riders that cross the minefield.
    var enemy = hero.findNearestEnemy();
    if (enemy) {
        if (hero.canCast("regen"))
            hero.cast("regen", hero);
        while (enemy.health > 0)
            hero.attack(enemy);
    }
}

while(true) {
    commandAttack();
    var strategy = chooseStrategy();
    // Call a function, depending on what the current strategy is.
    if (strategy == "griffin-rider")
        hero.summon("griffin-rider");
    else if (strategy == "fight-back")
        heroAttack();
    else {
        commandAttack();
        pickUpCoin();
    }
}
